File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/compoundviews/SplashAnimation.kt
package com.crescentflare.piratesgame.components.compoundviews

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import com.crescentflare.piratesgame.R
import com.crescentflare.piratesgame.components.basicviews.GradientView
import com.crescentflare.piratesgame.components.complexviews.PublisherLogo
import com.crescentflare.piratesgame.components.containers.FrameContainerView
import com.crescentflare.piratesgame.components.simpleviewlets.ImageViewlet
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.piratesgame.infrastructure.events.AppEvent
import com.crescentflare.piratesgame.infrastructure.events.AppEventObserver
import com.crescentflare.piratesgame.components.utility.ImageSource
import com.crescentflare.unilayout.helpers.UniLayoutParams
import com.crescentflare.unilayout.views.UniImageView
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.JsonLoader
import com.crescentflare.jsoninflator.binder.InflatableRef
import com.crescentflare.jsoninflator.binder.InflatorAnnotationBinder
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil

/**
 * Compound view: contains the publisher logo and background, handles animation
 */
class SplashAnimation : FrameContainerView {

    // --
    // Statics
    // --

    companion object {

        // --
        // Static: reference to layout resource
        // --

        const val layoutResource = R.raw.splash_animation


        // --
        // Static: viewlet integration
        // --

        val viewlet: JsonInflatable = object : JsonInflatable {

            override fun create(context: Context): Any {
                return SplashAnimation(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is SplashAnimation) {
                    // Apply background
                    obj.gradientColor = mapUtil.optionalColor(attributes, "gradientColor", 0)
                    obj.backgroundImage = ImageSource.fromObject(attributes["backgroundImage"])

                    // Apply state
                    obj.autoAnimation = mapUtil.optionalBoolean(attributes, "autoAnimation", false)
                    obj.on = mapUtil.optionalBoolean(attributes, "on", false)

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)

                    // Event handling
                    obj.setOnEvent = AppEvent.fromObject(attributes["setOnEvent"])

                    // Forward event observer
                    if (parent is AppEventObserver) {
                        obj.eventObserver = parent
                    }
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj is SplashAnimation
            }

        }

    }


    // --
    // Bound views
    // --

    @InflatableRef("logo")
    private var logoView: PublisherLogo? = null

    @InflatableRef("backgroundGradient")
    private var backgroundGradientView: GradientView? = null

    @InflatableRef("backgroundImage")
    private var backgroundImageView: UniImageView? = null


    // --
    // Members
    // --

    private var currentOn = false


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
            context: Context,
            attrs: AttributeSet? = null,
            defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
            context: Context,
            attrs: AttributeSet?,
            defStyleAttr: Int,
            defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        ViewletUtil.assertInflateOn(this, JsonLoader.instance.loadAttributes(context, layoutResource),null, InflatorAnnotationBinder(this))
        backgroundGradientView?.alpha = 0.0f
        clipChildren = false
    }


    // --
    // Configurable values
    // --

    var setOnEvent: AppEvent? = null

    var gradientColor: Int
        get() = backgroundGradientView?.endColor ?: 0
        set(gradientColor) {
            backgroundGradientView?.endColor = gradientColor
        }

    var backgroundImage: ImageSource? = null
        set(backgroundImage) {
            field = backgroundImage
            if (backgroundImageView?.drawable != null || currentOn) {
                ImageViewlet.applyImageSource(backgroundImageView, backgroundImage)
            }
        }

    var autoAnimation = false

    var on: Boolean
        get() = currentOn
        set(on) {
            setOn(on, autoAnimation)
        }

    fun setOn(on: Boolean, animated: Boolean) {
        // Safeguard against non-changing operations
        if (on == currentOn) {
            return
        }

        // Change state with optional animation
        if (animated) {
            logoView?.setOn(on, true, {
                ImageViewlet.applyImageSource(backgroundImageView, backgroundImage)
            }, {
                // Animate
                val animation = AnimatorSet()
                animation.playTogether(
                    ValueAnimator.ofFloat(if (on) 0.5f else 0.15f, if (on) 0.15f else 0.5f).apply {
                        addUpdateListener {
                            val value = it.animatedValue
                            if (value is Float) {
                                (logoView?.layoutParams as? UniLayoutParams)?.verticalGravity = value
                                logoView?.requestLayout()
                            }
                        }
                    },
                    ObjectAnimator.ofFloat(backgroundGradientView, View.ALPHA, if (on) 0f else 1f, if (on) 1f else 0f)
                )
                animation.duration = 500
                animation.startDelay = 500
                animation.interpolator = AccelerateDecelerateInterpolator()
                animation.start()

                // Apply state
                val setOnEvent = this.setOnEvent
                currentOn = on
                if (setOnEvent != null && on) {
                    eventObserver?.observedEvent(setOnEvent, this)
                }
            })
        } else {
            val setOnEvent = this.setOnEvent
            currentOn = on
            logoView?.setOn(on, false)
            if (on) {
                ImageViewlet.applyImageSource(backgroundImageView, backgroundImage)
            }
            backgroundGradientView?.alpha = if (on) 1.0f else 0.0f
            if (setOnEvent != null && on) {
                eventObserver?.observedEvent(setOnEvent, this)
            }
        }
    }

}

File: pirates-ios/PiratesGame/Components/CompoundViews/SplashAnimation.swift
//
//  SplashAnimation.swift
//  Compound view: contains the publisher logo and background, handles animation
//

import UIKit
import UniLayout
import JsonInflator

class SplashAnimation: FrameContainerView {

    // --
    // MARK: Layout JSON
    // --
    
    private let layoutFile = "SplashAnimation"


    // --
    // MARK: Bound views
    // --
    
    private var logoView: PublisherLogo?
    private var backgroundGradientView: GradientView?
    private var backgroundImageView: UniImageView?


    // --
    // MARK: Members
    // --

    private var currentOn = false

    
    // --
    // MARK: Viewlet integration
    // --
    
    override class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return SplashAnimation()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let animation = object as? SplashAnimation {
                // Apply background
                animation.gradientColor = convUtil.asColor(value: attributes["gradientColor"]) ?? UIColor.black
                animation.backgroundImage = ImageSource(string: convUtil.asString(value: attributes["backgroundImage"]))
                
                // Apply state
                animation.autoAnimation = convUtil.asBool(value: attributes["autoAnimation"]) ?? false
                animation.on = convUtil.asBool(value: attributes["on"]) ?? false
                
                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: animation, attributes: attributes)

                // Event handling
                animation.setOnEvent = AppEvent(value: attributes["setOnEvent"])

                // Forward event observer
                if let eventObserver = parent as? AppEventObserver {
                    animation.eventObserver = eventObserver
                }
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is SplashAnimation
        }
        
    }


    // --
    // MARK: Initialization
    // --

    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    fileprivate func setup() {
        // Bind views
        let binder = InflatorDictBinder()
        ViewletUtil.assertInflateOn(view: self, attributes: JsonLoader.shared.attributesFrom(jsonFile: layoutFile), parent: nil, binder: binder)
        logoView = binder.findByReference("logo") as? PublisherLogo
        backgroundGradientView = binder.findByReference("backgroundGradient") as? GradientView
        backgroundImageView = binder.findByReference("backgroundImage") as? UniImageView
        
        // Default properties
        backgroundGradientView?.alpha = 0
    }
    

    // --
    // MARK: Configurable values
    // --
    
    var setOnEvent: AppEvent?
    
    var gradientColor: UIColor {
        set {
            backgroundGradientView?.endColor = newValue
        }
        get { return backgroundGradientView?.endColor ?? UIColor.clear }
    }
    
    var backgroundImage: ImageSource? {
        didSet {
            if backgroundImageView?.image != nil || currentOn {
                ImageViewlet.applyImageSource(imageView: backgroundImageView, source: backgroundImage)
            }
        }
    }

    var autoAnimation = false

    var on: Bool {
        set {
            setOn(newValue, animated: autoAnimation)
        }
        get { return currentOn }
    }
    
    func setOn(_ on: Bool, animated: Bool) {
        // Safeguard against non-changing operations
        if on == currentOn {
            return
        }
        
        // Change state with optional animation
        if animated {
            logoView?.setOn(on, animated: true, afterAssetLoad: {
                ImageViewlet.applyImageSource(imageView: self.backgroundImageView, source: self.backgroundImage)
            }, completion: {
                UIView.animate(withDuration: 0.5, delay: 0.5, options: [.curveEaseInOut], animations: {
                    self.backgroundGradientView?.alpha = on ? 1 : 0
                    self.logoView?.transform = CGAffineTransform(translationX: 0, y: on ? -UIScreen.main.bounds.height / 6 : 0)
                })
                self.currentOn = on
                if let setOnEvent = self.setOnEvent, on {
                    self.eventObserver?.observedEvent(setOnEvent, sender: self)
                }
            })
        } else {
            currentOn = on
            logoView?.setOn(on, animated: false)
            if on {
                ImageViewlet.applyImageSource(imageView: backgroundImageView, source: backgroundImage)
            }
            backgroundGradientView?.alpha = on ? 1 : 0
            if let setOnEvent = setOnEvent, on {
                eventObserver?.observedEvent(setOnEvent, sender: self)
            }
        }
    }
   
}
